api/osb/store_instances_plugin.go

```diff
@@ -0,0 +1,707 @@
+	}
+	return nil
+}
+
+func (ssi *StoreServiceInstancePlugin) updateInstance(ctx context.Context, req *updateRequest, storage storage.Repository, usable bool) error {
+	byID := query.ByField(query.EqualsOperator, "id", req.InstanceID)
+	var instance types.Object
+	var err error
+	if instance, err = storage.Get(ctx, types.ServiceInstanceType, byID); err != nil {
+		if err != util.ErrNotFoundInStorage {
+			return util.HandleStorageError(err, string(types.ServiceInstanceType))
+		}
+	}
+	if instance == nil {
+		return nil
+	}
+	serviceInstance := instance.(*types.ServiceInstance)
+	oldServicePlanID := serviceInstance.ServicePlanID
+	previousValuesBytes, err := json.Marshal(req.PreviousValues)
+	if err != nil {
+		return err
+	}
+	previousValuesBytes, err = sjson.SetBytes(previousValuesBytes, smServicePlanIDKey, oldServicePlanID)
+	if err != nil {
+		return err
+	}
+	if len(req.PlanID) != 0 && req.PreviousValues.PlanID != req.PlanID {
+		var err error
+		serviceInstance.ServicePlanID, err = findServicePlanIDByCatalogIDs(ctx, storage, req.BrokerID, req.ServiceID, req.PlanID)
+		if err != nil {
+			return err
+		}
+	}
+	if len(req.MaintenanceInfo) != 0 {
+		serviceInstance.MaintenanceInfo = req.MaintenanceInfo
+	}
+	if len(req.RawContext) != 0 {
+		serviceInstance.Context = req.RawContext
+	}
+
+	serviceInstance.PreviousValues = previousValuesBytes
+	if _, err := storage.Update(ctx, serviceInstance, query.LabelChanges{}); err != nil {
+		return util.HandleStorageError(err, string(serviceInstance.GetType()))
+	}
+
+	return nil
+}
+
+func (ssi *StoreServiceInstancePlugin) rollbackInstance(ctx context.Context, req commonOSBRequest, storage storage.Repository, usable bool) error {
+	byID := query.ByField(query.EqualsOperator, "id", req.GetInstanceID())
+	var instance types.Object
+	var err error
+	if instance, err = storage.Get(ctx, types.ServiceInstanceType, byID); err != nil {
+		return util.HandleStorageError(err, string(types.ServiceInstanceType))
+	}
+	if instance == nil {
+		return nil
+	}
+	serviceInstance := instance.(*types.ServiceInstance)
+	serviceInstance.Usable = usable
+
+	if _, ok := req.(*lastOperationRequest); ok {
+		previousValues := serviceInstance.PreviousValues
+		oldCatalogPlanID := gjson.GetBytes(previousValues, smServicePlanIDKey).String()
+		if len(oldCatalogPlanID) != 0 {
+			serviceInstance.ServicePlanID = oldCatalogPlanID
+		}
+		oldMaintenanceInfo := gjson.GetBytes(previousValues, "maintenance_info").Raw
+		if len(oldMaintenanceInfo) != 0 {
+			serviceInstance.MaintenanceInfo = []byte(oldMaintenanceInfo)
+		}
+	}
+
+	if _, err := storage.Update(ctx, serviceInstance, query.LabelChanges{}); err != nil {
+		return util.HandleStorageError(err, string(serviceInstance.GetType()))
+	}
+
+	return nil
+}
+
+func (ssi *StoreServiceInstancePlugin) updateInstanceReady(ctx context.Context, req commonOSBRequest, storage storage.Repository) error {
+	byID := query.ByField(query.EqualsOperator, "id", req.GetInstanceID())
```

Comment:
Might be better to pass as a parameter directly the instance id instead of the whole `osbRequest`